stream-chat-android-client/src/main/java/io/getstream/chat/android/client/experimental/socket/State.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.client.experimental.socket
+
+import androidx.annotation.VisibleForTesting
+import io.getstream.chat.android.client.clientstate.DisconnectCause
+import io.getstream.chat.android.client.events.ConnectedEvent
+import io.getstream.chat.android.client.experimental.socket.ws.OkHttpWebSocket
+
+/**
+ * @startuml
+ * state SocketState {
+ *  state Connecting
+ *  state Connected
+ *  state Disconnecting
+ *  state Disconnected
+ *  state Destroyed
+ * }
+ *
+ * [*] --> Disconnected
+ * Disconnected --> Connecting : Lifecycle.Started
+ * Disconnected --> Destroyed : Lifecycle.Terminate
+ *
+ * Connecting --> Connected : WebSocket.OnConnectionOpened
+ * Connecting --> Disconnected: WebSocket.Terminate
+ *
+ * Connected --> Disconnecting: Lifecycle.Stopped
+ * Connected --> Connected : OnConnectedEventReceived
+ * Connected --> Destroyed : Lifecycle.Terminate
+ * Connected --> Disconnected: WebSocket.Terminate
+ *
+ * Disconnecting --> Disconnected: WebSocket.Terminate
+ *
+ * Destroyed --> [*]
+ * @enduml
+ */
+/**
+ * State of the socket connection.
+ */
+@VisibleForTesting
+internal sealed class State {
+
+    /**
+     * State of socket when connection is being established.
+     */
+    data class Connecting(val webSocket: OkHttpWebSocket) : State()
+
+    /**
+     * State of socket when the connection is established.
+     */
+    data class Connected(val event: ConnectedEvent?, val webSocket: OkHttpWebSocket) : State()
```

Comment:
I'm a bit concerned about having a `webSocket` instance as state property.

I think `webSocket` instance should be a part of `ChatSocket`

Reply:
Websocket instance lives only till we have the `Connecting` and `Connected` states. If we keep it in `ChatSocket`, it becomes our responsibility to clear it when the state changes or the web socket is not required anymore.

Reply:
I mean the `state` should not have a ref to the object which **state** it tries to represent.

Reply:
In other words, we should not expose `OkHttpWebSocket` instance by the published state hierarchy.

Reply:
This state machine reflects the state of `ChatSocket` which is not completely a representation of the `okhttp3.Websocket` state. I think of it as a `Session` that contains the `okhttp3.Websocket` instance used for that session.

Reply:
If `State` represents `ChatSocket` state, then why `State` exposes `ChatSocket's` internal implementation thing like `okhttp3.Websocket`?

Reply:
Ah, I see. Actually, the `State` is not supposed to be accessed outside `ChatSocket`. It is marked as `@VisibleForTesting`. I will refactor its usage outside this class.